code/game/machinery/mind_machine.dm

```diff
@@ -0,0 +1,846 @@
+	..()
+	if(istype(A, /obj/item/bluespace_crystal/flawless))
+		user.drop_item(A, src)
+		bluespaceConduit += 300
+		to_chat(user, "The bluespace conduit flashes violently before calming to a glow.")
+		qdel(A)
+		return
+	if(istype(A, /obj/item/bluespace_crystal/artificial))
+		user.drop_item(A, src)
+		bluespaceConduit += 1
+		to_chat(user, "The crystal is assimilated into the conduit!")
+		qdel(A)
+		return
+	if(istype(A, /obj/item/bluespace_crystal))
+		user.drop_item(A, src)
+		bluespaceConduit += 3
+		to_chat(user, "The crystal is assimilated into the conduit!")
+		qdel(A)
+		return
+	if(istype(A, /obj/item/device/soulstone))
+		if(soulShardSafety != FALSE)
+			to_chat(user, "That slot is full!")
+			return
+		if(A.contents.len)
+			to_chat(user, "The stone must be empty to function properly.") //prevents deleting players in stones
+			return
+		else
+			user.drop_item(A, src)
+			to_chat(user, "The stone is pulled into the machine. You can hear crunching and a quiet sizzle from inside.")
+			qdel(A)
+			soulShardSafety = TRUE
+
+/obj/machinery/mind_machine/mind_machine_hub/attack_hand(user as mob)
+	if(podsConnected == FALSE || connectOne == FALSE || connectTwo == FALSE)
+		podsConnected = FALSE
+		connectOne = FALSE
+		connectTwo = FALSE
+		findConnections() //No UI without pods
+	if(!podsConnected)
+		to_chat(user, "Pod connection error.")
+		return
+	ui_interact(user)
+
+//////UI stuff///////////////
+
+/obj/machinery/mind_machine/mind_machine_hub/ui_interact(mob/user, ui_key = "main", var/datum/nanoui/ui = null, var/force_open = NANOUI_FOCUS)
+	var/list/data = list(
+			"nameOne" = occupantNameOne,
+			"nameTwo" = occupantNameTwo,
+			"statOne" = occupantStatOne,
+			"statTwo" = occupantStatTwo,
+			"mindTypeOne" = mindTypeOne,
+			"mindTypeTwo" = mindTypeTwo,
+			"lockedPods" = lockedPods,
+			"bluespaceConduit" = bluespaceConduit,
+			"soulShardSafety" = soulShardSafety,
+			"manipRating" = manipRating,
+			"swapProgress" = swapProgress,
+			"errorMessage" = errorMessage
+	)
+
+	ui = nanomanager.try_update_ui(user, src, ui_key, ui, data, force_open)
+
+	if(!ui)
+		ui = new(user, src, ui_key, "mind_machine.tmpl", "Mind Machine", 350,500)
+		ui.set_initial_data(data)
+		ui.open()
+	ui.set_auto_update(1)
+
+/obj/machinery/mind_machine/mind_machine_hub/Topic(href, href_list)
+	if(href_list["scan_start"])
+		if(!currentlySwapping)
+			scanOccupants()
+			return TRUE
+		else
+			errorMessage = "Currently active: Please wait"
+			return TRUE
+	if(href_list["unlock_pods"])
+		if(((!connectOne.contents.len) || (!connectTwo.contents.len)) && (currentlySwapping))
+			unlockPods() //Safeguard against the pod getting trapped in infinite scan
+			return TRUE
+		if(lockedPods == FALSE)
+			errorMessage = "Pods already unlocked"
+			return TRUE
+		if(currentlySwapping)
+			errorMessage = "Currently active: Please wait"
+			return TRUE
+		else
+			spawn(5)
```

Comment:
Returning from spawn like this doesn't work correctly. Either way you should definitely have player feedback that the button *did* something (instead of basically causing artificial UI lag).

If you still want to have this take a bit of time you can call `nanomanager.update_uis(src)` to force update the UI externally.